openmmtools/integrators.py

```diff
@@ -1895,6 +1905,145 @@ def _add_alchemical_reset_step(self):
         # Add all dependent parameters
         self._add_update_alchemical_parameters_step()
 
+class PeriodicNonequilibriumIntegrator(AlchemicalNonequilibriumLangevinIntegrator):
+    """
+    Periodic nonequilibrium integrator where master alchemical parameter ``lambda``
+    is driven through a periodic protocol:
+
+    eq 0 for nsteps_eq | neq 0->1 over nsteps_neq | eq 1 for nsteps_eq | neq 1->0 for nsteps_neq
+
+    Arbitrary Langevin splitting strings can be provided.
+
+    .. warning :: This API is experimental, and subject to change.
+
+    Examples
+    --------
+
+    Create a nonequilibrium integrator to switch the center of a harmonic oscillator between two locations, dwelling in either state to allow equilibration.
+    This example uses BAOAB (V R O R V) for integration.
+
+    >>> # Create harmonic oscillator testsystem
+    >>> from openmmtools import testsystems
+    >>> from simtk import openmm, unit
+    >>> testsystem = testsystems.HarmonicOscillator()
+    >>> # Create a nonequilibrium alchemical integrator
+    >>> alchemical_functions = { 'testsystems_HarmonicOscillator_x0' : 'lambda' }
+    >>> nsteps_eq = 100 # number of steps to dwell in equilibrium at lambda = 0 or 1
+    >>> nsteps_neq = 50 # number of steps in the switching trajectory where lambda is switched from 0 to 1
+    >>> integrator = PeriodicAlchemicalNonequilibriumLangevinIntegrator(
+    ...                  temperature=300*unit.kelvin, collision_rate=1.0/unit.picoseconds, timestep=1.0*unit.femtoseconds,
+    ...                  alchemical_functions=alchemical_functions, splitting="V R H O R V",
+    ...                  nsteps_eq=nsteps_eq, nsteps_neq=nsteps_neq)
+    >>> # Create a Context
+    >>> context = openmm.Context(testsystem.system, integrator)
+    >>> # Run one periodic cycle: (eq 0) > (neq 0->1) > (eq 1) > (neq 1->0)
+    >>> context.setPositions(testsystem.positions)
+    >>> nsteps_per_period = 2*nsteps_eq + 2*nsteps_neq
+    >>> integrator.step(nsteps_per_period)
```

Comment:
it seems like this class was written to allow for single `integrator.step()` calls which don't complete the nonequilibrium _cycle_; is this correct?

Reply:
That's correct! It's the same way that `AlchemicalNonequilibriumLangevinIntegrator` works, except instead of halting after `n_steps_neq`, it continues periodic cycling infinitely. This way, we can also measure the work at various points through the switching, and this allows us to capture useful information about the efficiency of the protocol.

I think your integrators are super useful as well, so I'll come back to those in a couple of days.